File: DSA/Java/src/dsa/lib/algo/sort/MergeSort.java
package dsa.lib.algo.sort;

import dsa.lib.Utils;

public class MergeSort {
    public void sort(int[] ary, int n) {
        split(ary, 0, n - 1);
    }

    /**
     * Split the array into 2 sub-arrays, then merge the 2 sub-arrays into a sorted array.
     *
     * @param ary The array to be split.
     * @param st  The start position of the array to be split, inclusive.
     * @param end The end position of the array to be split, inclusive.
     */
    private void split(int[] ary, int st, int end) {
        if (st >= end) return;

        int mid = st + (end - st) / 2;
        split(ary, st, mid);
        split(ary, mid + 1, end);
        merge(ary, st, mid, end);

    }

    /**
     * Merge 2 sub-arrays into a sorted array.
     *
     * @param ary The given array contains 2 sub-arrays.
     * @param st  The start position of the 1st sub-array, inclusive.
     * @param m   The end position of the 1st sub-array, inclusive.
     *            Also it is the divided position of the 2 sub-arrays, the next position should be the start of the 2nd sub-array.
     * @param end The end position of the 2nd sub-array, inclusive.
     */
    private void merge(int[] ary, int st, int m, int end) {
        final int n = end - st + 1;
        int[] result = new int[n];

        for (int i = st, j = m + 1, k = 0; k < n; ++k) {
            if (i <= m && j <= end) {
                if (ary[i] <= ary[j]) result[k] = ary[i++];
                else result[k] = ary[j++];
            } else if (i <= m) {
                result[k] = ary[i++];
            } else result[k] = ary[j++];
        }

        Utils.arraycopy(result, 0, ary, st, end - st + 1);
    }
}

File: DSA/Python/src/dsa/lib/algo/sort/merge_sort.py
from typing import List

from dsa.lib.utils import array_copy


class MergeSort:
    def sort(self, ary: List[int], n: int) -> None:
        self.split(ary, 0, n - 1)

    def split(self, ary: List[int], st: int, end: int) -> None:
        """
        Split the array into 2 sub-arrays recursively, then merge them into a sorted array.

        :param ary:
        :param st: The start position of the array to be split, inclusive.
        :param end: The end position of the array to be split, inclusive.
        """
        if st >= end: return

        mid: int = st + (end - st) // 2
        self.split(ary, st, mid)
        self.split(ary, mid + 1, end)
        self.merge(ary, st, mid, end)

    @staticmethod
    def merge(ary: List[int], st: int, m: int, end: int) -> None:
        """
        Merge 2 sub-arrays into a sorted array.

        :param ary:
        :param st: The start position of 1st sub-array, inclusive.
        :param m: The end position of the 1st sub-array, inclusive.
        Also it is the divided position of the 2 sub-arrays, the next position is the start of the 2nd sub-array.
        :param end: The end position of the 2nd array,inclusive.
        :return:
        """
        n: int = end - st + 1
        result: List[int] = [0 for _ in range(n)]

        i, j = st, m + 1
        for k in range(n):
            if i <= m and j <= end:
                if ary[i] <= ary[j]:
                    result[k] = ary[i]
                    i += 1
                else:
                    result[k] = ary[j]
                    j += 1
            elif i <= m:
                result[k] = ary[i]
                i += 1
            else:
                result[k] = ary[j]
                j += 1

        array_copy(result, 0, ary, st, n)
